File: yujung/복습/미로탈출.cpp
#include <string>
#include <vector>
#include<queue>
using namespace std;

char map[100][100];
int n;
int m;
int dx[4]={0,0,1,-1};
int dy[4]={1,-1,0,0};
int sx,sy;
int lx,ly;
int ex,ey;
bool is_range(int x,int y){
    return x>=0&&y>=0&&x<n&&y<m;
}
int bfs(int x,int y,int ex,int ey){
    int cnt=0;
    bool visited[100][100]={false};
    queue<pair<int,pair<int,int>>> q;
    q.push({x,{y,cnt}});
    
    visited[x][y]=true;
    while(!q.empty()){
        auto a=q.front();
        q.pop();
        int cx=a.first;
        int cy=a.second.first;
        int ccount=a.second.second;
        if(cx==ex&&cy==ey){return ccount;}
        for(int i=0; i<4;i++){
            int nx=cx+dx[i];
            int ny=cy+dy[i];
            if(!is_range(nx,ny)||visited[nx][ny]==true||map[nx][ny]=='X'){
                continue;
            }
            
            q.push({nx,{ny,ccount+1}});
            visited[nx][ny]=true;
            
        }
    }
    return -1;
}
int solution(vector<string> maps) {
    int answer = 0;
     n=maps.size();
     m=maps[0].size();
    for(int i=0; i<n;i++){
        for(int j=0; j<m;j++){
            map[i][j]=maps[i][j];
            if(map[i][j]=='S'){
                sx=i;
                sy=j;
            }else if(map[i][j]=='L'){
                lx=i;
                ly=j;
            }
            else if(map[i][j]=='E'){
                ex=i;
                ey=j;
            }
        }
    }
    
    int sum1=bfs(sx,sy,lx,ly);

    if(sum1==-1){
        return -1;
    }else{
        int sum2=bfs(lx,ly,ex,ey);
        if(sum2==-1){
            return -1;
        }else{
            return sum1+sum2;
        }
    }
  
}

File: seoyeon/프로그래머스/연습문제/미로 탈출.py
#정확성: 91.3

from collections import deque

dp = []
n = 0
m = 0

dx = [0,0,-1,1]
dy = [-1,1,0,0]

def bfs(maps,sx,sy,check):
    visited=[[False for _ in range(m)] for _ in range(n)]
    Q = deque()
    Q.append((sx,sy))
    visited[sx][sy] = True
    if check == 0:
        match = "L"
    else:
        match = "E"
    
    while Q:
        x,y = Q.popleft()
        for d in range(4):
            nx = x+dx[d]
            ny = y+dy[d]
            if 0<=nx<n and 0<=ny<m:
                if visited[nx][ny]==False and maps[nx][ny] == "O": 
                    dp[nx][ny] = dp[x][y] + 1
                    visited[nx][ny] = True
                    Q.append((nx,ny))
                if visited[nx][ny]==False and maps[nx][ny] == match:
                    dp[nx][ny] = dp[x][y] + 1
                    visited[nx][ny] = True
                    return dp[nx][ny]
    
    return -1
    


def solution(maps):
    global n,m,dp
    answer,check = 0,0
    n = len(maps)
    m = len(maps[0])
    
    dp = [[0 for _ in range(m)] for _ in range(n)]
    s,e,l = [-1,-1], [-1,-1], [-1,-1]
    
    for i in range(n):
        for j in range(m):
            if maps[i][j] == "S":
                s = [i,j]
            if maps[i][j] == "L":
                l = [i,j]
            if maps[i][j] == "E":
                e = [i,j]
    
    #BFS 사용 => 레버 탐색
    answer = bfs(maps,s[0],s[1],check)
    if answer == -1:
        return answer
    check = 1
    #BFS 사용 => 출구 탐색
    answer = bfs(maps,l[0],l[1],check)
    
    return answer

#정확성 100/100
from collections import deque

dp = []
n = 0
m = 0

dx = [0,0,-1,1]
dy = [-1,1,0,0]

def bfs(maps,sx,sy,check):
    visited=[[False for _ in range(m)] for _ in range(n)]
    Q = deque()
    Q.append((sx,sy))
    visited[sx][sy] = True
    if check == 0:
        match = "L"
    else:
        match = "E"
    
    while Q:
        x,y = Q.popleft()
        for d in range(4):
            nx = x+dx[d]
            ny = y+dy[d]
            if 0<=nx<n and 0<=ny<m:
                if visited[nx][ny]==False and maps[nx][ny] == match:
                    dp[nx][ny] = dp[x][y] + 1
                    visited[nx][ny] = True
                    return dp[nx][ny]
                
                if visited[nx][ny]==False and maps[nx][ny] != "X": 
                    dp[nx][ny] = dp[x][y] + 1
                    visited[nx][ny] = True
                    Q.append((nx,ny))

    return -1
    


def solution(maps):
    global n,m,dp
    answer,check = 0,0
    n = len(maps)
    m = len(maps[0])
    
    dp = [[0 for _ in range(m)] for _ in range(n)]
    s,e,l = [-1,-1], [-1,-1], [-1,-1]
    
    for i in range(n):
        for j in range(m):
            if maps[i][j] == "S":
                s = [i,j]
            if maps[i][j] == "L":
                l = [i,j]
            if maps[i][j] == "E":
                e = [i,j]
    
    #BFS 사용 => 레버 탐색
    answer = bfs(maps,s[0],s[1],check)
    if answer == -1:
        return -1
    check = 1
    #BFS 사용 => 출구 탐색
    answer = bfs(maps,l[0],l[1],check)
    
    return answer
